ncclient/transport/ssh.py

```diff
@@ -371,25 +389,57 @@ def connect(self, host, port=830, timeout=None, unknown_host_cb=default_unknown_
                 sock = socket.fromfd(int(sock_fd), socket.AF_INET, socket.SOCK_STREAM)
             sock.settimeout(timeout)
 
-        t = self._transport = paramiko.Transport(sock)
-        t.set_log_channel(logger.name)
+        self._transport = paramiko.Transport(sock)
+        self._transport.set_log_channel(logger.name)
         if config.get("compression") == 'yes':
-            t.use_compression()
-
+            self._transport.use_compression()
+
+        # Set preferred host keys, to those we possess for the host
+        # Otherwise may have a situation where known_hosts contains a valid key for the host,
+        # but that key is not selected during negotiation
+        if self._host_keys:
+            if port == PORT_SSH_DEFAULT:
+                known_hosts_lookup = host
+            else:
+                known_hosts_lookup = '[%s]:%s' % (host, port)
+            known_host_keys_for_this_host = self._host_keys.lookup(known_hosts_lookup)
+            if known_host_keys_for_this_host:
+                self._transport._preferred_keys = [x.key.get_name() for x in known_host_keys_for_this_host._entries]
+
+        # If we need to connect with a specific hostkey, negotiate for only its type
+        if hostkey:
+            hostkey_obj = None
+            for key_cls in [paramiko.DSSKey, paramiko.Ed25519Key, paramiko.RSAKey, paramiko.ECDSAKey]:
+                try:
+                    hostkey_obj = key_cls(data=base64.b64decode(hostkey))
+                except paramiko.SSHException as e:
+                    # Not a key of this type - try the next
+                    pass
+            if not hostkey_obj:
+                # We've tried all known host key types and haven't found which one to use - bail
+                raise e
+            self._transport._preferred_keys = [hostkey_obj.get_name()]
```

Comment:
If we have both `self._host_keys` and the `hostkey` parameter, we set `self._transport._preferred_keys` once above and then re-set it again here, losing the previously set value. That feels incorrect to me. Should the first `if` statement be something like `if self._host_keys and not hostkey:` instead?

Reply:
The first 'if' is for when host keys have been set by code above (i.e. ssh_config and we read known_hosts)

However regardless of this - if the hostkey has been specified - it takes precedence.